aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,973 @@
+
+        let target_size = ((tree.order as u64) + 1) / 2;
+
+        let l = 0;
+        let r = current_size;
+        while (l != r) {
+            let mid = l + (r - l) / 2;
+            if (vector::borrow(children, mid).max_key < key) {
+                l = mid + 1;
+            } else {
+                r = mid;
+            };
+        };
+
+        let left_node_index = bucket_table::length(&tree.nodes) + 2;
+
+        if (parent_index == NULL_INDEX) {
+            // Splitting root now, need to create a new root.
+            parent_index = bucket_table::length(&tree.nodes) + 3;
+            node.parent = parent_index;
+
+            tree.root_index = parent_index;
+            let parent_node = new_node(/*is_leaf=*/false, /*parent=*/NULL_INDEX);
+            let max_element = vector::borrow(children, current_size - 1).max_key;
+            if (max_element < key) {
+                max_element = key;
+            };
+            vector::push_back(&mut parent_node.children, new_child(max_element, node_index));
+            bucket_table::add(&mut tree.nodes, parent_index, parent_node);
+        };
+
+        let right_node = new_node(*is_leaf, parent_index);
+
+        right_node.next = *next;
+        *next = node_index;
+        right_node.prev = left_node_index;
+        if (*prev != NULL_INDEX) {
+            bucket_table::borrow_mut(&mut tree.nodes, *prev).next = left_node_index;
+        };
+
+        if (l < target_size) {
+            let i = target_size - 1;
+            while (i < current_size) {
+                vector::push_back(&mut right_node.children, *vector::borrow(children, i));
+                i = i + 1;
+            };
+
+            while (current_size > target_size) {
+                vector::pop_back(children);
+                current_size = current_size - 1;
+            };
+
+            i = target_size - 1;
+            while (i > l) {
+                *vector::borrow_mut(children, i) = *vector::borrow(children, i - 1);
+                i = i - 1;
+            };
+            *vector::borrow_mut(children, l) = child;
+        } else {
+            let i = target_size;
+            while (i < l) {
+                vector::push_back(&mut right_node.children, *vector::borrow(children, i));
+                i = i + 1;
+            };
+            vector::push_back(&mut right_node.children, child);
+            while (i < current_size) {
+                vector::push_back(&mut right_node.children, *vector::borrow(children, i));
+                i = i + 1;
+            };
+
+            while (current_size > target_size) {
+                vector::pop_back(children);
+                current_size = current_size - 1;
+            };
+        };
+
+        if (!*is_leaf) {
+            let i = 0;
+            while (i < target_size) {
+                bucket_table::borrow_mut(&mut tree.nodes, vector::borrow(children, i).node_index).parent = left_node_index;
```

Comment:
This will end up being quote costly, depending on how many children each node has and how well things are grouped in the bucket table. Ignoring per byte costs, each read costs at least 3000 octas and each write costs at least 3000 octas.

One way to get around this is avoid having a parent reference at each of the nodes. For each split, you don't need to worry about updating the many children.

This'll make iterating harder but you can get around this by pushing all values to the leafs and connecting each successive leaf (like a B+ tree). Tradeoff there is more nodes overall but it's still asymptotically O(log(n))

Reply:
Thanks for the feedback. You are right this could potentially cost more gas.

(To clarify, this is a B+tree implementation and all values are on the leaf)

I store parent reference mainly because:
0. I do the recursion for insertion/deletion bottom-up.
1. For deletion, when merging with the brother node, I need to know whether the next node at the same level share the same parent with the node-to-be-merged.

I was assuming these split/merge operation won't happen very frequently, so the amortized cost won't be huge. In addition to that, in a foreseeable future we might want to change our gas schedule to reduce the cost of any operation that doesn't have permanent effect. (e.g. read/write existing item)

I'd like to do more analysis on the gas side on real use cases with the help from @lightmark , before deciding if I should make change here. Theoretically I think this (parent) information doesn't have to be persistent, I can keep it on stack and pass around if necessary.

Reply:
Makes sense.

> change our gas schedule to reduce the cost of any operation that doesn't have permanent effect. (e.g. read/write existing item)

^ This would be tremendously helpful! Per item and per byte costs to read and write existing items in global storage end up being quite costly.